oak-store-document/src/test/java/org/apache/jackrabbit/oak/plugins/document/WriteAfterRecoveryTest.java

```diff
@@ -0,0 +1,661 @@
+        assertEquals(lateWriteRevisions, inconsistent);
+    }
+
+    @Test
+    public void orphanedNodesAdded() throws Exception {
+        merge(ns1, createChild("/a"), true);
+
+        List<UpdateOp> failed = new ArrayList<>();
+        store.addListener(filter(failed, ADD_NODE_OPS));
+        // let merge create some nodes
+        store.fail().after(2).eternally();
+        try {
+            merge(ns1, createChild("/a/b/c/d/e/f"), false);
+            fail("merge must fail");
+        } catch (CommitFailedException e) {
+            // expected
+        }
+        disposeQuietly(ns1);
+        store.fail().never();
+
+        // wait until lease expires
+        clockWait(DEFAULT_LEASE_DURATION_MILLIS);
+
+        DocumentNodeStore ns2 = createNodeStore(2);
+        assertTrue(ns2.getLastRevRecoveryAgent().isRecoveryNeeded());
+        assertEquals(1, ns2.getLastRevRecoveryAgent().recover(1));
+
+        // 'late write'
+        store.createOrUpdate(NODES, failed);
+        Set<Revision> lateWriteRevisions = getRevisions(failed);
+
+        // revive clusterId 1
+        ns1 = createNodeStore(1);
+        merge(ns1, createChild("/d"), true);
+
+        ns2.runBackgroundOperations();
+        assertTrue(getNode(ns2.getRoot(), "/d").exists());
+        assertFalse(getNode(ns2.getRoot(), "/a/b").exists());
+        // can create nodes that recovery removed ...
+        merge(ns2, createChild("/a/b/c"), false);
+        // ... but not those written after recovery
+        try {
+            merge(ns2, createChild("/a/b/c/d/e/f"), false);
+            fail("merge must fail");
+        } catch (CommitFailedException e) {
+            assertThat(e.getMessage(), containsString("older than base"));
+        }
+
+        Set<Revision> inconsistent = checkConsistency(ns2, "/a/b", "/a/b/c", "/a/b/c/d", "/a/b/c/d/e", "/a/b/c/d/e/f");
+        assertEquals(lateWriteRevisions, inconsistent);
+    }
+
+    @Test
+    public void oneNodeRemoved() throws Exception {
+        merge(ns1, createChild("/a/b"), true);
+
+        List<UpdateOp> failed = new ArrayList<>();
+        store.addListener(filter(failed, REMOVE_NODE_OPS));
+        store.fail().after(0).eternally();
+        try {
+            merge(ns1, removeChild("/a/b"), false);
+            fail("merge must fail");
+        } catch (CommitFailedException e) {
+            // expected
+        }
+        disposeQuietly(ns1);
+        store.fail().never();
+
+        // wait until lease expires
+        clockWait(DEFAULT_LEASE_DURATION_MILLIS);
+
+        DocumentNodeStore ns2 = createNodeStore(2);
+        assertTrue(ns2.getLastRevRecoveryAgent().isRecoveryNeeded());
+        assertEquals(0, ns2.getLastRevRecoveryAgent().recover(1));
+
+        // 'late write'
+        store.createOrUpdate(NODES, failed);
+        Set<Revision> lateWriteRevisions = getRevisions(failed);
+
+        // revive clusterId 1
+        ns1 = createNodeStore(1);
+        merge(ns1, createChild("/d"), true);
```

Comment:
```suggestion
        assertEquals(1, checkConsistency(ns1, "/a/b").size());
        merge(ns1, createChild("/d"), true);
        assertEquals(0, checkConsistency(ns2, "/a/b").size());
```

Might be useful to add for illustration, that `ns1` already sees the inconsistency, while `ns2` requires a backgroundRead.

Reply:
Good suggestion. I added a slightly different assert, but functionally what you suggested.